README.md
# jestlearn
## serverless website for interactive educational content
[![Netlify Status](https://api.netlify.com/api/v1/badges/4250a53b-cc48-482c-ae06-c07cbffd75f6/deploy-status)](https://app.netlify.com/sites/jestlearn/deploys)
Make multichoice/input problems with ease!

## How to use:
`data.ts` is the active file read by the rest of the program, so just copy and paste your course data into it and generate the `public` folder and distribute that! (note: typescript/something will try to auto-fix imports! make sure every component is only reading `./data` and not `./data_computer_systems`, etc)
Edit `course_name` in `data.ts`, should be unique
Edit `data.ts` with problem sets and such and that's mostly it
If you want cloud syncing change the `website_sync_point` and self host the server. For now it's pointed at my server
If you want to support more languages, just replace prism.js and prism.css

## Legend:
✅ - for correct answer
❌ - for incorrect answer
⚠️ - for corrected incorrect answer
❓ - for unanswered

## Todo:
* Anki spaced alg.
* consider removing choice problem because all select problems are choice problems, but not all choice problems are select ones, maybe there's a way to limit the checkbox selection
* pure offline support
* Add more documentation, make it easier to set up a sync server, a video to show how to set it up
* GUI editor for problems? (too lazy atm, just edit js)
* Analytics page for teachers
* provide a way to add comments other than github comments?
* better device scaling support, use tachyons or something.. my css is terrible 

## Motivation:
< 1000 lines of code
I just want something super simple, fast, flexible(open source), and offline to make small educational exercise content. Google classroom and everything else is too bloat, not to mention I don't really care about grades. Instead of focusing on grades, focus on constant repitition, ergo Anki, because memory is fragile.

Mistranslated quote but emphasis on doing always rather than listening/watching
> What I hear, I forget; What I see, I remember; What I do, I understand. —Confucius, 551–479 BC
> What I cannot create I do not understand - Feynman
## Credits:
* Favicon provided by https://twemoji.twitter.com/


CHANGELOG.md
# 0.6.1
- Critical Fix bug where problems would not generate at all
+ Fixed semantic versioning

# 0.0.6
- Switched to giscus using github discussion feature instead of issues(turns out utterances doesn't work at all due to re-direct)
- Using giscus to store and review student submissions for now
- TODO: when github discussions introduce templates, use those to force correct submission structure
- Added code submission/discussion problems

# 0.0.5
- Now titles are used for comparisons rather than ids since ids are hard to make unique and are tied to section-ids(maybe crc32 the titles for uids)
- Prevent re-doing of problems that are already marked correct when the user did not click "Next" to proceed.
- Added clear cache button
- Added show answer button
- Added cloudflare analytics
- Input answers are now space insensitive
- Input answer hints now can take html
- Breaking ProblemSet constructor because search tags are kinda optional and should be last
# 0.0.4
- Fixed TOC not updating
- Merged back other deleted props rather than just .gen()
- Page no longer refreshes after logging in
- Added support for hitting enter key to proceed for InputProblems
- No longer using + - ! ? , all emoji now
# 0.0.3
- No longer saving problems that have not been worked on in localstorage
- Fixed diffing algorithim to merge ones that dont exist along with getting rid of duplicates
- Fixed issues with the discussions page for single domain courses
- Added hover effect to select answers
- Added async support to fetch problems from an api and an example on how to do so
- Removed num of problems from the problem set in favor closures that hold that privately since it was not used anywhere else
- Fixed input problems not supporting syntax highlighting

src/utils.ts
import { COURSE_NAME } from "./data";
import type { ProblemSet } from "./ProblemSet";

///replace spaces with underscores and lower cases the name
export function convert_to_hash(st: string): string {
  let regex = / /gi;
  let replace_spaces = st.replace(regex, "_");
  let res = replace_spaces.toLowerCase();
  return res;
}

export enum Sync {
  UPDATE = 0,
  INITIAL = 1,
  ARCHIVE = 2,
}

export function diff_latest(set_a: ProblemSet[], set_b: ProblemSet[]) {
  let bigger = set_a;
  let smaller = set_b;
  if (set_a.length < set_b.length) {
    bigger = set_b;
    smaller = set_a;
  }
  //merge the unique ones that the other doesnt have
  for (const item of smaller) {
    let is_unique = true;
    for (const item2 of bigger) {
      if (item.title === item2.title) {
        is_unique = false;
        break;
      }
    }
    if (is_unique) {
      bigger.push(item);
    }
  }
  //diffing
  let diffed = bigger.map((val) => {
    let keep = val;
    for (const val2 of smaller) {
      if (val.title == val2.title && val.last_updated < val2.last_updated) {
        keep = val2;
        break;
      }
    }
    return keep;
  });
  return diffed;
}

const website_sync_point = "https://jestlearn.com/sync";
//const website_sync_point = "http://localhost:3000/sync";
export async function send_sync(username, TOC, server_copy = [], code = 0):Promise<any> {
  let dlatest = diff_latest(TOC, server_copy);
  let send = {
    code: code,
    course_name: COURSE_NAME,
    username: username,
    problem_sets: dlatest
      .filter((val, index) => {
        let serv = server_copy.find((i) => i.title === val.title);
        let update_server_sync = false;
        if (serv) {
          update_server_sync = serv.last_updated < val.last_updated;
        } else if (val.last_updated !== 0) {
          //does not exist on the server yet, but only send if they edited it
          update_server_sync = true;
        }
        //only sync when server data is older than then one here on the client
        return update_server_sync;
      })
      .map((pset) => {
        //no need to send tags& resources, mainly just problem data:
        //have to deep clone
        let clone = JSON.parse(JSON.stringify(pset));
        delete clone.resources;
        delete clone.tags;
        clone.problems = clone.problems.map((problem) => {
          // keep input_answer_hints and explanations because they might be procedurally generated explanations
          //delete problem.input_answer_hint;
          if (Array.isArray(problem.answer)) {
            problem.answer = problem.answer.map((answ) => {
              //delete answ.explanation;
              return answ;
            });
          }
          return problem;
        });
        return clone;
      }),
  };
  ///console.log(send);
  let response = await fetch(website_sync_point, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(send),
  });
  return response.json();
}


src/ProblemSet.ts
export class Problem {
  type: string;
  question: string;
  answer: string | number | { text: string; correct: boolean }[]; //todo: should probably be an array of strings for multi-answer input
  input_answer_hint: string;
  result: string;
  tries: number;
  time: number;
  hints: number;
  constructor(
    type: string,
    question: string,
    answer:
      | string
      | number
      | { text: string; correct: boolean; explanation: string }[],
    input_answer_hint: string = ""
  ) {
    this.type = type;
    this.question = question;
    this.answer = answer;
    this.input_answer_hint = input_answer_hint;
    this.result = "❓";
    this.tries = 0;
    this.time = 0;
    this.hints = 0;
  }
}

export class ProblemSet {
  title: string;
  ///id is for section identification
  id: number;
  problem_index: number;
  tags: string[];
  // if you are not randomly generating problems, stick them in .problems
  problems: Problem[];
  //otherwise you are, so provide a function that returns an array of problems, or an async function if you need to fetch from an api
  gen: Function | undefined;
  resources: { url_title: string; url: string; additional: string }[]; //ray of links to theoretical content, additional can be some html though its not recommended because its better to have things open in a new tab rather than needing to keep popping the modal on and off
  emoji_mark: string;
  last_updated: number;
  constructor(
    title: string,
    id: number,
    gen: Function | undefined,
    problems: Problem[] = [],
    resources: { url_title: string; url: string; additional: string }[] = [],
    tags: string[] = [],
    emoji_mark: string = "❓",
  ) {
    this.title = title;
    this.id = id;
    this.problem_index = 0;
    this.problems = problems;
    this.tags = tags;
    this.gen = gen;
    this.resources = resources;
    this.emoji_mark = emoji_mark;
    this.last_updated = 0;
  }
}
export function has_started(problems: Problem[]): boolean {
  for (let i = 0; i < problems.length; i++) {
    if (problems[i].tries >= 1) {
      return true;
    }
  }
  return false;
}

//adds back the gen() functions that were not seralized
export function merge_back_deleted_props(without: ProblemSet[], original: ProblemSet[]): ProblemSet[] {
  return without.map((val) => {
    let found = original.find((hay) => val.title === hay.title);
    if(found) {
      val.gen = found.gen;
      val.resources = found.resources;
      val.tags = found.tags;
      return val;
    } else {
      return val;
    }
  })
}

export function gen_amount(amount = 1, f: () => Problem): () => Problem[] {
  return function g() {
    let res = [];
    for(let i = 0; i < amount; i++){
      res.push(f());
    }
    return res;
  };
}


src/data_example.ts
import { ProblemSet, Problem } from "./ProblemSet";

let e_choice = new Problem(
  "select",
  "Which of the following statements are true about computers",
  [
    { text: "computers use 0 and 1s", correct: true, explanation: "" },
    {
      text: "computers were invented in the stone era",
      correct: false,
      explanation: "",
    },
    { text: "computers use electricity", correct: true, explanation: "" },
  ],
  "hinty"
);

let e_choice2 = new Problem(
  "select",
  "Which of the following statements are false about water?",
  [
    {
      text: "water is abbreviated h2o in chemistry",
      correct: false,
      explanation: "<i>Yes but that is not what the question is asking</i>",
    },
    {
      text: "water is not necssary for human survival",
      correct: true,
      explanation: "",
    },
    {
      text: "most plants do not need water to survive",
      correct: true,
      explanation: "",
    },
  ],
  "hinty"
);

let e_choice_code = new Problem(
  "select",
  `<pre class="line-numbers match-braces rainbow-braces"><code class="language-racket">
; What is the next stemp in the evaluation order?: 
(+ 1 3 (* 50 3))
</code></pre>`,
  [
    {
      text: `<pre class="line-numbers match-braces rainbow-braces"><code class="language-racket">
(+ 1 3 (* 150))
</code></pre>`,
      correct: true,
      explanation:
        "Good job! We look to the most inward parens and evaluate that one first.",
    },
    {
      text: `<pre class="line-numbers match-braces rainbow-braces"><code class="language-racket">
(+ 4 (* 50 3))
</code></pre>`,
      correct: false,
      explanation:
        "from left to right, find the most inward parens and evaluate it",
    },
  ]
);

let example_multi_choice = new ProblemSet(
  "Example Multi Choice",
  1,
  undefined,
  [e_choice, e_choice2, e_choice_code],
  [
    { url_title: "📄 articles", url: "https://google.com", additional: "" },
    { url_title: "🎥 video", url: "https://youtube.com", additional: "" },
  ],
  ["tag1", "tag2"],
);

// min and max are inclusive
function ran_int(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function ran_bool() {
  return (ran_int(0, 1) == 1);
}

let example_generate = new ProblemSet(
  "Example Input RanGen: Binary To Decimal",
  2.0,
  function generate_binary_to_decimal(amount = 5) {
    function normalize_bwidth(num, width) {
      // zero shift right converts the number to a signed 32 bit number?
      let binary = (num >>> 0).toString(2);
      if (binary.length < width) {
        //pad with 0s
        return binary.padStart(width, "0");
      } else if (binary.length > width) {
        return binary.slice(-width);
      } else if (binary.length == width) {
        return binary;
      } else {
        console.warn(
          "WARNING! CHECK TO MAKE SURE THE BIT-WIDTH FITS WITHIN NUM RANGE"
        );
        return binary;
      }
    }
    function subscript(string, base) {
      return `${string}<sub>${base}</sub>`;
    }
    let num = ran_int(0, 255);
    let binary = normalize_bwidth(num, 8);
    return Array(amount).fill(new Problem(
      "input",
      `<h2>Convert ${subscript(binary, 2)} to decimal?</h2>`,
      parseInt(binary, 2).toString(),
      "Answer ex: 3"
    ));
  },
  [],
  [
    {
      url_title: "article",
      url: "https://computers404.netlify.app/06-binarynumbers",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ],
  ["additionalsearcheterms", "convert"],
);

let example_rest_api = new ProblemSet(
  "Racket Expressions",
  3.0,
  async function get_it() {
    let id = 0;
    let amount = 5;
    let response = await fetch("http://jestlearn.com/exercise", {
      method: "POST",
      headers: {
      "Content-Type": "application/json",
      },
      body: JSON.stringify({"id": id, "amount": amount})
    });
    let result = await response.json();
    return result;
  },
  [],
  [
    {
      url_title: "article",
      url: "https://computers404.netlify.app/06-binarynumbers",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ],
  ["additionalsearcheterms", "convert"],
);

function pick_random_el(arr: any[]) {
  return arr[Math.floor(Math.random() * arr.length)];
}
let all: ProblemSet[] = [
  example_multi_choice,
  example_generate,
  example_rest_api,
];
//WARNING: do not stringify this! we need the gen function and a copy of the original questions for resetting and other things
export const TOC_original: ProblemSet[] = all.map((val) => Object.freeze(val));

// we are mutating this
//copy over the gen() functions that got removed because of stringify
export let TOC: ProblemSet[] = JSON.parse(JSON.stringify(all)).map(
  (val, index) => {
    let m1 = Object.assign(val, TOC_original[index]);
    //so we can run has_started() since it deleted the prototype link to the clas :/
    m1.__proto__ = ProblemSet.prototype;
    return m1;
  }
);
//window.toc = TOC;
//the home page of your course will be at https://yourdomain.com/example_course , where example_course is the name of the course(all spaces are replaced by underscores) 
export const COURSE_NAME = "example course";

src/data_howtocode.ts
import { object_without_properties } from "svelte/internal";
import { ProblemSet, Problem, gen_amount } from "./ProblemSet";

function fetch_problems(id: number = 0, amount = 5) {
return async function get_it() {
    //const fetch_url = "http://localhost:3000/exercise";
    const fetch_url = "https://jestlearn.com/exercise";
    let response = await fetch(fetch_url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: id, amount: amount }),
    });
    let result = await response.json();
    return result;
  }
}

/// min and max are inclusive
function ran_int(min = 0, max = 1) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function ran_bool() {
  return ran_int(0, 1) == 1;
}

let racket_math_expressions = new ProblemSet(
  "Expressions and Evaluation with numbers",
  2.0,
  fetch_problems(0, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let infix_to_prefix = new ProblemSet(
  "Translate infix to prefix math",
  2.1,
  fetch_problems(2.1, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let racket_string_practice = new ProblemSet(
  "Zero based indexing string practice",
  2.2,
  gen_amount(10, function gen_problem() {
    let ran_words = [
      "StRiKeBrEaKeR",
      "InVeStIgAtIoN",
      "AlLoCaTiOn",
      "EnViRoNmEnT",
      "CoRrEcTiOn",
      "CoRrEsPoNdEnCe",
      "InVeStMeNt",
      "DeMoCrAtIc",
      "CoNcLuSiOn",
      "MaInStReAm",
      "PoSsIbIlItY",
      "AdMiRaTiOn",
      "InTeLlIgEnCe",
      "DiSaBiLiTy",
      "ExCiTeMeNt",
      "ReAsOnAbLe",
      "InStRuCtIoN",
      "DiScIpLiNe",
      "NeGoTiAtIoN",
      "OpErAtIoNaL",
    ];
    function ran_caps(arr: string[]) {
      return arr.map(function random_caps(val) {
        return val
          .split("")
          .map((letter, index) => {
            if (index % 2 == 0) {
              return letter.toUpperCase();
            } else {
              return letter;
            }
          })
          .reduce((acc, curr) => acc + curr);
      });
    }
    let word: string = pick_random_el(ran_words);
    let start = ran_int(0, word.length - 1);
    let end = ran_int(start, word.length - 1);

    //the answer has to be quoted on the offchance that the string is empty
    let result = new Problem(
      "input",
      `<pre class="line-numbers match-braces rainbow-braces"><code class="language-racket">; what does the following produce?
(substring "${word}" ${start} ${end})</code></pre>`,
      `"${word.substring(start, end)}"`,
      'remember: caps matter, wrap your answer in quotes! e.g: "bOo"'
    );
    return result;
  }),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/strings.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let ran_var_arith = new ProblemSet(
  "Variable arith evaluation",
  2.3,
  fetch_problems(2.3, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let build_a_string = new ProblemSet(
  "Build a string",
  2.4,
  fetch_problems(2.4, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let comparisons = new ProblemSet(
  "If bool comparisons",
  2.5,
  fetch_problems(2.5, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let logical = new ProblemSet(
  "If logical",
  2.6,
  fetch_problems(2.6, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

function pick_random_el(arr: any[]) {
  return arr[Math.floor(Math.random() * arr.length)];
}
let all: ProblemSet[] = [racket_math_expressions, infix_to_prefix, racket_string_practice, ran_var_arith, build_a_string, comparisons, logical];
//WARNING: do not stringify this! we need the gen function and a copy of the original questions for resetting and other things
export const TOC_original: ProblemSet[] = all.map((val) => Object.freeze(val));

// we are mutating this
//copy over the gen() functions that got removed because of stringify
export let TOC: ProblemSet[] = JSON.parse(JSON.stringify(all)).map(
  (val, index) => {
    let m1 = Object.assign(val, TOC_original[index]);
    //so we can run has_started() since it deleted the prototype link to the clas :/
    m1.__proto__ = ProblemSet.prototype;
    return m1;
  }
);
//window.toc = TOC;
//the home page of your course will be at https://yourdomain.com/example_course , where example_course is the name of the course(all spaces are replaced by underscores)
export const COURSE_NAME = "how to code"

src/data.ts
import { object_without_properties } from "svelte/internal";
import { ProblemSet, Problem, gen_amount } from "./ProblemSet";

function fetch_problems(id: number = 0, amount = 5) {
return async function get_it() {
    //const fetch_url = "http://localhost:3000/exercise";
    const fetch_url = "https://jestlearn.com/exercise";
    let response = await fetch(fetch_url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: id, amount: amount }),
    });
    let result = await response.json();
    return result;
  }
}

/// min and max are inclusive
function ran_int(min = 0, max = 1) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function ran_bool() {
  return ran_int(0, 1) == 1;
}

let racket_math_expressions = new ProblemSet(
  "Expressions and Evaluation with numbers",
  2.0,
  fetch_problems(0, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://howtocode.pages.dev/expressions",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/nBUcisbrvTo", additional: "" },
  ]
);

let infix_to_prefix = new ProblemSet(
  "Translate infix to prefix math",
  2.1,
  fetch_problems(2.1, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/Nk2mpEgc2F0", additional: "" },
  ]
);

let racket_string_practice = new ProblemSet(
  "Zero based indexing string practice",
  2.2,
  gen_amount(10, function gen_problem() {
    let ran_words = [
      "StRiKeBrEaKeR",
      "InVeStIgAtIoN",
      "AlLoCaTiOn",
      "EnViRoNmEnT",
      "CoRrEcTiOn",
      "CoRrEsPoNdEnCe",
      "InVeStMeNt",
      "DeMoCrAtIc",
      "CoNcLuSiOn",
      "MaInStReAm",
      "PoSsIbIlItY",
      "AdMiRaTiOn",
      "InTeLlIgEnCe",
      "DiSaBiLiTy",
      "ExCiTeMeNt",
      "ReAsOnAbLe",
      "InStRuCtIoN",
      "DiScIpLiNe",
      "NeGoTiAtIoN",
      "OpErAtIoNaL",
    ];
    function ran_caps(arr: string[]) {
      return arr.map(function random_caps(val) {
        return val
          .split("")
          .map((letter, index) => {
            if (index % 2 == 0) {
              return letter.toUpperCase();
            } else {
              return letter;
            }
          })
          .reduce((acc, curr) => acc + curr);
      });
    }
    let word: string = pick_random_el(ran_words);
    let start = ran_int(0, word.length - 1);
    let end = ran_int(start, word.length - 1);

    //the answer has to be quoted on the offchance that the string is empty
    let result = new Problem(
      "input",
      `<pre class="line-numbers match-braces rainbow-braces"><code class="language-racket">; what does the following produce?
(substring "${word}" ${start} ${end})</code></pre>`,
      `"${word.substring(start, end)}"`,
      'remember: caps matter, wrap your answer in quotes! e.g: "bOo"'
    );
    return result;
  }),
  [],
  [
    {
      url_title: "article",
      url: "https://howtocode.pages.dev/strings#zero_based_indexing",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/oddmn_fD_vU", additional: "" },
  ]
);

let ran_var_arith = new ProblemSet(
  "Variable arith evaluation",
  2.3,
  fetch_problems(2.3, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://howtocode.pages.dev/variables",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/WViW7y17UhE", additional: "" },
  ]
);

let build_a_string = new ProblemSet(
  "Build a string",
  2.4,
  fetch_problems(2.4, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://howtocode.pages.dev/variables",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bGdl11zoRlc", additional: "" },
  ]
);

let comparisons = new ProblemSet(
  "If bool comparisons",
  2.5,
  fetch_problems(2.5, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://howtocode.pages.dev/booleans",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/_25wDPmuCKc", additional: "" },
  ]
);

let logical = new ProblemSet(
  "If logical",
  2.6,
  fetch_problems(2.6, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://howtocode.pages.dev/booleans",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/O8c4arZ0FW8", additional: "" },
  ]
);

let func_scope_eval = new ProblemSet(
  "Function Evaluation and Scope 1",
  2.7,
  fetch_problems(2.7, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let func_scope_eval_2 = new ProblemSet(
  "Function Evaluation and Scope 2 nested",
  2.8,
  fetch_problems(2.8, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

let structs_build = new ProblemSet(
  "Structs",
  2.9,
  fetch_problems(2.9, 10),
  [],
  [
    {
      url_title: "article",
      url: "https://jesthowtocode.netlify.app/expressions.html",
      additional: "",
    },
    { url_title: "video", url: "https://youtu.be/bFLB4dyNKUk", additional: "" },
  ]
);

function pick_random_el(arr: any[]) {
  return arr[Math.floor(Math.random() * arr.length)];
}

// files and their names stored at : https://github.com/jestarray/howtocode/starter
function exercise_from_filenames() {
  // TODO: make this an object with youtube links
  let filenames = [
    "4-string-first_starter.rkt",
    "4.1-string-last_starter.rkt",
    "4.2-string-rest_starter.rkt",
    "4.3-string-remove-last_starter.rkt",
    "4.4-formalize_starter.rkt",
    "5-image-area_starter.rkt",
    "5.1-red-x_starter.rkt",
    "5.2-cancel_starter.rkt",
    "6-ensure-question_starter.rkt",
    "6.1-punctuate_starter.rkt",
  ];
  return filenames.map((item) => {
    let dash_index = item.indexOf("-");
    let problem_number = parseFloat(item.slice(0, dash_index));
    let problem_name = item.slice(dash_index + 1).replace("_starter.rkt", "");
    return new ProblemSet(problem_name, problem_number, undefined, [new Problem("submission", 
    `<h1>⬇️<a href="https://howtocode.pages.dev/starter/${item}" download>Download the starter file</a></h1>`, "")], 
    [{url_title: "Solution", url: `https://howtocode.pages.dev/solutions/${item.replace("_starter.rkt", "_solution.rkt")}`, additional: ""}])
  });
}

let all: ProblemSet[] = [racket_math_expressions, infix_to_prefix, racket_string_practice, ran_var_arith, build_a_string, comparisons, logical, func_scope_eval, func_scope_eval_2, structs_build, ...exercise_from_filenames()];
//WARNING: do not stringify this! we need the gen function and a copy of the original questions for resetting and other things
export const TOC_original: ProblemSet[] = all.map((val) => Object.freeze(val));

// we are mutating this
//copy over the gen() functions that got removed because of stringify
export let TOC: ProblemSet[] = JSON.parse(JSON.stringify(all)).map(
  (val, index) => {
    let m1 = Object.assign(val, TOC_original[index]);
    //so we can run has_started() since it deleted the prototype link to the clas :/
    m1.__proto__ = ProblemSet.prototype;
    return m1;
  }
);
//window.toc = TOC;
//the home page of your course will be at https://yourdomain.com/example_course , where example_course is the name of the course(all spaces are replaced by underscores)
export const COURSE_NAME = "how to code"